src/unity/python/turicreate/test/test_activity_classifier.py

```diff
@@ -17,28 +17,59 @@
 from . import util as test_util
 import pytest
 from turicreate.toolkits._internal_utils import _mac_ver
+import uuid
 
 
-def _load_data(self):
+def _load_data(self, num_examples = 1000, num_features = 3, max_num_sessions = 4,
+               randomize_num_sessions = True, num_labels = 9, prediction_window = 5,
+               enforce_all_sessions = False):
     random.seed(42)
 
-    self.num_examples = 1000
-    self.num_features = 3
-    self.num_sessions = random.randint(1, 4)
-    self.num_labels = 9
-    self.prediction_window = 5
+    self.num_examples = num_examples
+    self.num_features = num_features
+    self.num_sessions = random.randint(1, max_num_sessions) if randomize_num_sessions else max_num_sessions
+    self.num_labels = num_labels
+    self.prediction_window = prediction_window
 
     self.features = ['X1-r', 'X2-r', 'X3-r']
     self.target = 'activity_label'
     self.session_id = 'session_id'
 
-    random_session_ids = sorted([random.randint(0, self.num_sessions - 1) for i in range(self.num_examples)])
+    if (enforce_all_sessions):
+        random_session_ids = _random_session_ids(self.num_examples, self.num_sessions)
+    else:
+        random_session_ids = sorted([random.randint(0, self.num_sessions - 1) for i in range(self.num_examples)])
+
     random_labels = [random.randint(0, self.num_labels - 1) for i in range(self.num_examples)]
 
     self.data = tc.util.generate_random_sframe(column_codes='r' * self.num_features, num_rows=self.num_examples)
-    self.data['session_id'] = random_session_ids
+    self.data[self.session_id] = random_session_ids
     self.data[self.target] = random_labels
 
+'''
+    Creates a random session_id column, that guarantees that the number
+    of sessions is exactly the requested one.
+'''
+def _random_session_ids(num_examples , num_sessions):
+    examples_per_session = num_examples // num_sessions
+    if (examples_per_session == 0):
+        raise ValueError("Can't divide {} lines into {} sessions.".format(num_examples, num_sessions))
+
+    min_lines_per_session = int(0.85 * examples_per_session)
+    max_lines_per_session = int(1.15 * examples_per_session)
+
+    lines_in_each_session = [random.randint(min_lines_per_session, max_lines_per_session) for i in range(num_sessions)]
+    lines_in_each_session[-1] += num_examples - sum(lines_in_each_session)
+
+    if lines_in_each_session[-1] < 0:
+        raise ValueError()
```

Comment:
This doesn't guard against `lines_in_each_session[-1] == 0`

And the probability of an error increases as the number of sessions increases.

It might be more robust just to normalize the sampled `lines_in_each_session`, by multiplying each value by `num_examples / sum(lines_in_each_session)`, followed by the existing fudge factor (in line 62) to deal with rounding.

(Interestingly, this sampling procedure would begin to resemble how you sample multinomial parameters from a Dirichlet distribution: https://en.wikipedia.org/wiki/Dirichlet_distribution#Random_number_generation)

Reply:
I've known my way of guaranteeing `sum(lines_in_each_session) == num_examples` is off right when I've written it - but couldn't think of a better method. Yours is the better (and probably the correct) method. Thanks!

I'll read about the Dirichlet sampling tomorrow, when my mind is still fresh. Right now the very mention of it brings back haunting memories from horrible college courses... 😄

Reply:
Eh, the link I posted is of mostly academic interest. The sampling procedure boils down to generating `num_sessions` draws from a random distribution and then normalizing to obtain a random list of `num_sessions` probabilities that sum to 1.

The Dirichlet connection comes in that you can view `lines_in_each_session` as a multinomial random variable that we could sample e.g. using https://en.wikipedia.org/wiki/Multinomial_distribution#Sampling_from_a_multinomial_distribution. But we don't really care about `lines_in_each_session` actually having a multinomial distribution. Sampling a Dirichlet is a convenient way of giving us something similar, literally choosing a random multinomial distribution (represented by its parameters).

Maybe the real question is to what degree the test even relies on `lines_in_each_session` being random....

Reply:
None of our tests are truly random at the moment.  
They all set the seed to `42` (always!), and most don't even use randomness.  

Nevertheless - I do believe that we'll add proper pseudo-randomness to the tests someday - so it's better to write new tests with that mindset.